sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -0,0 +1,244 @@
+# Minimum spanning tree single linkage implementation for hdbscan
+# Authors: Leland McInnes <leland.mcinnes@gmail.com>
+#          Steve Astels <sastels@gmail.com>
+#          Meekail Zain <zainmeekail@gmail.com>
+# License: 3-clause BSD
+
+cimport numpy as cnp
+from libc.float cimport DBL_MAX
+
+import numpy as np
+from ...metrics._dist_metrics cimport DistanceMetric
+from ...cluster._hierarchical_fast cimport UnionFind
+from ...cluster._hdbscan._tree cimport HIERARCHY_t
+from ...cluster._hdbscan._tree import HIERARCHY_dtype
+from ...utils._typedefs cimport intp_t, float64_t, int64_t, uint8_t
+
+cdef extern from "numpy/arrayobject.h":
+    ctypedef struct PyArrayObject
```

Comment:
[`PyArrayObject` is listed in NumPy Cython API as of Cython 0.29.X](https://github.com/cython/cython/blob/8af954f812f78d5258cb8440b13cc6944eccf940/Cython/Includes/numpy/__init__.pxd#L237-L240) (but `PyArray_SHAPE` isn't).

Can `PyArrayObject` be `cimport`ed from `cnp`?

This suggestion applies to other files.